src/main/java/mate/academy/EventManager.java

```diff
@@ -1,19 +1,45 @@
 package mate.academy;
 
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
 public class EventManager {
-    public void registerListener(EventListener listener) {
+    private CopyOnWriteArrayList<EventListener> events = new CopyOnWriteArrayList<>();
+    private final ExecutorService executor;
 
+    public EventManager() {
+        this.events = new CopyOnWriteArrayList<>();
+        this.executor = Executors.newCachedThreadPool();
     }
 
-    public void deregisterListener(EventListener listener) {
+    public void registerListener(EventListener listener) {
+        events.add(listener);
+    }
 
+    public void deregisterListener(EventListener listener) {
+        events.remove(listener);
     }
 
     public void notifyEvent(Event event) {
-
+        for (EventListener listener : events) {
+            executor.submit(() -> listener.onEvent(event));
+        }
     }
 
     public void shutdown() {
-
+        executor.shutdown();
+        try {
+            if (!executor.awaitTermination(60, TimeUnit.SECONDS)) {
+                executor.shutdownNow();
+                if (!executor.awaitTermination(60, TimeUnit.SECONDS)) {
+                    System.err.println("Executor did not terminate");
+                }
+            }
+        } catch (InterruptedException ie) {
+            executor.shutdownNow();
+            Thread.currentThread().interrupt();
```

Comment:
Catching 'InterruptedException' and calling 'Thread.currentThread().interrupt()' is a good practice as it allows other methods higher up the call stack to check the interrupted status of the thread. However, make sure that this behavior is documented and that any calling code is prepared to handle the interrupted status appropriately.